typescript/infra/src/utils/utils.ts

```diff
@@ -240,3 +240,39 @@ export function streamToString(stream: NodeJS.ReadableStream): Promise<string> {
       .on('end', () => resolve(String.prototype.concat(...chunks)));
   });
 }
+
+/**
+ * Converts a matrix to 1d array ordered by diagonals. This is useful if you
+ * want to make sure that the order operations are performed in are ordered but
+ * not repeating the same values from the inner or outer array in sequence.
+ *
+ * @warn Requires a square matrix.
+ *
+ * // 0,0 1,0 2,0 3,0
+ * //
+ * // 0,1 1,1 2,1 3,1
+ * //
+ * // 0,2 1,2 2,2 3,2
+ * //
+ * // 0,3 1,3 2,3 3,3
+ *
+ * becomes
+ *
+ * 0,0; 1,0; 0,1; 2,0; 1,1; 0,2; 3,0; 2,1; 1,2; 0,3; 3,1; 2,2; 1,3; 3,2; 2,3; 3,3
```

Comment:
why is this ordering best?
at a glance, repeating destination 0 twice in a row feels unideal

Reply:
Well, it's a lot better than just going though in a `for origin in origins { for destination in origins.filter(...) { ... } }` and I don't have a better idea off the top of my head short of shuffling but that would potentially be a worse order AND it would be unpredictable.

So if you have an ordering in mind I am all ears.